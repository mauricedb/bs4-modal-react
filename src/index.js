export * from './modal';
export * from './modal-header';
export * from './modal-title';
export * from './modal-body';
