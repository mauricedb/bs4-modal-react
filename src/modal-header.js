import React, { Component } from "react";
import PropTypes from "prop-types";

export class ModalHeader extends Component {
  static propTypes = {
    onHide: PropTypes.func.isRequired,
    closeButton: PropTypes.bool
  };

  static defaultProps = {
    closeButton: true
  };

  getCloseButtonElement() {
    const { closeButton, onHide } = this.props;

    if (!closeButton) {
      return null;
    }

    return (
      <button
        type="button"
        className="close"
        aria-label="Close"
        onClick={onHide}
      >
        <span aria-hidden="true">&times;</span>
      </button>
    );
  }
  render() {
    const { children } = this.props;
    const closeButton = this.getCloseButtonElement();

    return (
      <div className="modal-header">
        {children}
        {closeButton}
      </div>
    );
  }
}
