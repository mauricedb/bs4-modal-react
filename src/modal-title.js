import React from "react";

export const ModalTitle = ({ children }) =>
  <h5 className="modal-title">
    {children}
  </h5>;
