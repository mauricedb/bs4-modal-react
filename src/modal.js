import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import { ModalHeader } from "./modal-header";

const TRANSITION_DURATION = 300;

const ClassName = {
  MODAL: "modal",
  SCROLLBAR_MEASURER: "modal-scrollbar-measure",
  BACKDROP: "modal-backdrop",
  OPEN: "modal-open",
  FADE: "fade",
  SHOW: "show"
};

const Display = {
  HIDDEN: "hidden",
  SHOWING: "showing",
  VISIBLE: "visible",
  HIDING: "hiding"
};

export class Modal extends Component {
  static propTypes = {
    visible: PropTypes.bool.isRequired,
    onHide: PropTypes.func.isRequired,
    backdrop: PropTypes.bool,
    transition: PropTypes.bool
  };

  static defaultProps = {
    backdrop: true,
    transition: true
  };

  state = {
    handle: 0
  };

  clearTimer() {
    if (this.handle) {
      clearTimeout(this.handle);
      this.setState({ handle: 0 });
    }
  }

  getDisplay(props) {
    let display = "";
    const { visible, transition } = props;
    const animating = !!this.state.handle;

    if (visible) {
      if (animating) {
        display = Display.SHOWING;
      } else {
        display = Display.VISIBLE;
      }
    } else {
      if (animating) {
        display = Display.HIDING;
      } else {
        display = Display.HIDDEN;
      }
    }

    return display;
  }

  onClickModal = e => {
    if (e.target.classList.contains(ClassName.MODAL)) {
      this.props.onHide();
    }
  };

  componentWillReceiveProps(nextProps) {
    const display = this.getDisplay(nextProps);
    const { transition } = nextProps;

    switch (display) {
      case Display.HIDDEN:
        if (transition) {
          const handle = setTimeout(() => {
            this.setState({ handle: 0 });
          }, TRANSITION_DURATION);
          this.setState({ handle });
        }
        break;
      case Display.SHOWING:
        break;
      case Display.VISIBLE:
        if (transition) {
          const handle = setTimeout(() => {
            this.setState({ handle: 0 });
          }, TRANSITION_DURATION);
          this.setState({ handle });
        }
        break;
      case Display.HIDING:
        break;
    }
  }

  render() {
    const display = this.getDisplay(this.props);
    const { backdrop, transition, onHide, children } = this.props;

    if (display === Display.HIDDEN) {
      return null;
    }
    let backdropEl = null;

    if (backdrop) {
      backdropEl = (
        <div
          className={classnames(ClassName.BACKDROP, {
            [ClassName.FADE]: transition,
            [ClassName.SHOW]: display === Display.VISIBLE
          })}
        />
      );
    }

    return (
      <div>
        {backdropEl}
        <div
          className={classnames(ClassName.MODAL, {
            [ClassName.FADE]: transition,
            [ClassName.SHOW]: display === Display.VISIBLE
          })}
          style={{ display: "block" }}
          tabIndex="-1"
          role="dialog"
          onClick={this.onClickModal}
        >
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              {React.Children.map(children, child => {
                if (child.type === ModalHeader) {
                  return React.cloneElement(child, { onHide });
                }
                return child;
              })}
            </div>
          </div>
        </div>
      </div>
    );
  }
}
