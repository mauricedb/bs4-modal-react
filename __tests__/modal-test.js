import React from 'react';
import renderer from 'react-test-renderer';
import { Modal, ModalHeader, ModalTitle, ModalBody, ModalFooter } from "../src";

it('renders when invisible', () => {
  const tree = renderer.create(
      <Modal visible={false} onHide={() => {}}>
        <ModalHeader>
          <ModalTitle>Modal title</ModalTitle>
        </ModalHeader>
        <ModalBody>Hello Modal</ModalBody>
        <ModalFooter>
          <button type="button" className="btn btn-secondary">
            Close
          </button>
          <button type="button" className="btn btn-primary">
            Save changes
          </button>
        </ModalFooter>
      </Modal>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders when visible', () => {
  const tree = renderer.create(
      <Modal visible={true} onHide={() => {}}>
        <ModalHeader>
          <ModalTitle>Modal title</ModalTitle>
        </ModalHeader>
        <ModalBody>Hello Modal</ModalBody>
        <ModalFooter>
          <button type="button" className="btn btn-secondary">
            Close
          </button>
          <button type="button" className="btn btn-primary">
            Save changes
          </button>
        </ModalFooter>
      </Modal>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});