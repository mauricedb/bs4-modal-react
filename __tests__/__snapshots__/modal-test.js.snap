// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when invisible 1`] = `null`;

exports[`renders when visible 1`] = `
<div>
  <div
    className="modal-backdrop fade show"
  />
  <div
    className="modal fade show"
    onClick={[Function]}
    onKeyUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            Modal title
          </h5>
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          Hello Modal
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            type="button"
          >
            Close
          </button>
          <button
            className="btn btn-primary"
            type="button"
          >
            Save changes
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
